flink-runtime/src/main/java/org/apache/flink/runtime/rest/FileUploadHandler.java

```diff
@@ -116,5 +136,16 @@ private void reset() {
 		currentHttpPostRequestDecoder.destroy();
 		currentHttpPostRequestDecoder = null;
 		currentHttpRequest = null;
+		currentUploadDir = null;
+		currentJsonPayload = null;
+	}
+
+	public static Optional<byte[]> getMultipartJsonPayload(ChannelHandlerContext ctx) {
+		return Optional.ofNullable(ctx.channel().attr(UPLOADED_JSON).get());
+	}
+
+	public static FileUploads getMultipartFileUploads(ChannelHandlerContext ctx) {
+		return Optional.ofNullable(ctx.channel().attr(UPLOADED_FILES).get())
+			.orElse(FileUploads.EMPTY);
```

Comment:
How files are stored is an implementation detail of the `FileUploadHandler`, why would we expose this to subsequent handlers?

Reply:
True, then I would suggest to only store the directory in the `FileUploads` and adding a method to retrieve all uploaded files.

Reply:
How does this differ to the current implementation? Are you suggesting to simplify `FileUploads` to only consider the case of 1 directory containing N files?

Reply:
I think not much. My gut feeling is just that `FileUploads` can be simplified. Instead of having our own FileVisitor, we could simply call `FileUtils.deleteDirectory(uploadDirectory)`. And I think  this class has actually two responsibilities: Listing all files to make them accessible and storing the directories in which they reside to delete them afterwards.